Exception refactoring

Deraen opened the pull request:
We decided to drop support for old options completely. Options are validated in compile time so there shouldn't be surprising runtime problems.

The default exception handlers are now defined using `compojure.api.exception` namespace instead of `meta` or `middleware`. `coerce!` still throws `:ring.swagger.schema/validation` which we map to our type. This is probably temporary solution as `coerce!` will be moved to another project and when that is done we should probably revisit request validation exception throwing.

Supersedes #158 and #143

ikitommi commented on src/compojure/api/middleware.clj:
```diff
@@ -197,13 +194,15 @@
         {:keys [formats params-opts response-opts]} format]
+    ; Break at compile time if there are deprecated options
+    (assert (not (:error-handler (:validation-errors options))) "Deprecated option: [:validation-errors :error-handler], use [:exceptions :error-handlers :compojure.api.middleware/request-validation] instead.")
+    (assert (not (:catch-core-errors? (:validation-errors options))) "Deprecated option: [:validation-errors :catch-core-errors?], use [:exceptions :error-handlers :compojure.api.exception/request-validation] instead.")
```
this should be :schema.core/error?

ikitommi commented on src/compojure/api/middleware.clj:
```diff
@@ -171,11 +166,13 @@
-   - **:validation-errors**         for *ring.swagger.middleware/wrap-validation-errors*
-       - **:error-handler**           function to handle ring-swagger schema exceptions
-       - **:catch-core-errors?**      whether to catch also `:schema.core/errors`
+       - **:error-handlers**          map of error handlers for different error types.
```
`:error-handlers` could be just `:handlers`? as there is already the [Error](http://docs.oracle.com/javase/7/docs/api/java/lang/Error.html), which != Exception.

ikitommi commented on src/compojure/api/meta.clj:
```diff
@@ -59,7 +61,7 @@
           (let [body (schema/coerce schema (:body response) matcher)]
             (if (schema/error? body)
-              (internal-server-error {:errors (:error body)})
+              (throw+ (assoc body :type ::ex/response-validation))
```
src-coerce! in `meta.clj` should also use plain `coerce` and throw `::ex/request-validation`.

ikitommi commented on src/compojure/api/middleware.clj:
```diff
@@ -129,19 +133,10 @@
-                            :message (.getMessage e)})))
+(defn handle-req-error [^Throwable e handler request]
+  (if (or (instance? JsonParseException e) (instance? ParserException e))
+    (throw+ {:type ::ex/request-validation} e)
```
won't this break as `::ex/request-validation` handlers think it's a Schema Error? I guess there should be a separate type for these.

jkytomak replied:
New type means also new handler config to override for users.
Anyhow now the original exception type and message is missed

ikitommi commented:
Few comments, but looks great!

jkytomak commented on test/compojure/api/integration_test.clj:
```diff
@@ -204,8 +218,7 @@
         status => 400
-        (:type body) => "json-parse-exception"
-        (:message body) => truthy))))
+        (:errors body) => contains "Unexpected character"))))
```
I had typo there, check does not work. Should be:
(contains "Unexpected character")

jkytomak commented:
Overall looks good, some comments above. I tried changes also with our application.
Check it, if you are interested from user side:
https://github.com/Opetushallitus/valtionavustus/commit/6d9b20518f7a9d389b0e4b0132167a4f0300d347

In my version I had the error message body generating functions public so i could use same formatting as original. Now I had to copy paste a bit code to our codebase, but not a big deal.

Deraen commented:
Good comments.

I'm thinking changing the parameters of error handlers a bit:
Now the first parameter is either the Exception or attached map. Because of this sometimes handler won't have access to the original exception.

So instead of `[ex-or-data type req]` I'm thinking `[ex data req]`, as the type would be contained in data.

jkytomak commented:
Good idea. Notice that when calling deprecated handler, you have to try to choose right type of parameter for it.

ikitommi commented:
where do the response-parsing errors go? is the mw-pipeline set up in such order that new exception handler really catches the serialization exceptions too? test would be nice, I can do when this gets merged.

Deraen commented:
Response is serialized/encoded, not parsed :)

Both response and request parsing exceptions are a bit of a mess because r-m-f doesn't relly catch just parsing or serialization exceptions but ALL exceptions (i.e. `(handler req)` is inside try-catch).

Deraen commented:
R-m-f by default rethrows all the exceptions, request parsing is currently set up so that we rethrow selected exceptions (json and yaml) with correct type. We could implement the same for response serialization, but would be better if this can be fixed in r-m-f.

Deraen commented:
Still thinking what is the best way to handle old error-handlers and parsing and serialization errors, but besides those I think most of the changes proposed in comments should be implemented.

For old error-handlers I can see following three options:

1. Catching ArityExceptions when calling handlers
2. Checking arities of handlers beforehand (reflection or similar)
3. Mention changed arity in deprecated option errors

The last option would be  the simplest but is it enough?

jkytomak commented:
4. option would be catch errors in error handlers and safely handle them (be aware of infite loops)

Deraen commented:
I added back `call-error-handler` to catch ArityExceptions and call error-handler with one argument and to display a warning. This still can break old error-handlers which except to receive the data map instead of the exception.

I also added changelog entry about the changes, hopefully most of users will check that when upgrading.

Deraen commented:
Merged. If there are still something to improve, we can do that in master.

ikitommi commented:
Looks great!! README still needs updating on this. Volunteer(s)?
